pkg/controller/core.oam.dev/v1alpha2/application/generator.go

```diff
@@ -358,15 +508,24 @@ func (h *AppHandler) applyComponentFunc(appParser *appfile.Parser, appRev *v1bet
 	}
 }
 
-// overrideTraits will override cluster field to be local for traits which are control plane only
+// overrideTraits will override field for traits by TraitDefinition.Spec
 func overrideTraits(appRev *v1beta1.ApplicationRevision, readyTraits []*unstructured.Unstructured) []*unstructured.Unstructured {
 	traits := readyTraits
 	for index, readyTrait := range readyTraits {
 		for _, trait := range appRev.Spec.TraitDefinitions {
-			if trait.Spec.ControlPlaneOnly && trait.Name == readyTrait.GetLabels()[oam.TraitTypeLabel] {
+			if trait.Name != readyTrait.GetLabels()[oam.TraitTypeLabel] {
+				continue
+			}
+
+			stageType := trait.Spec.Stage
+			if len(stageType) == 0 {
+				stageType = v1beta1.DefaultDispatch
+			}
+			util.AddAnnotations(traits[index], map[string]string{oam.AnnotationTraitStageType: string(stageType)})
+
+			if trait.Spec.ControlPlaneOnly {
 				oam.SetCluster(traits[index], "local")
 				traits[index].SetNamespace(appRev.GetNamespace())
-				break
```

Comment:
why omit the break ? \cc @wangyikewxgm

Reply:
Type judgment has been added before this.

Reply:
I  think `break` and `continue` are different logic, the `break` will execute only once in the loop while the continue means we can execute more than once if condition matches. Maybe we still need the `break` here.